src/shm/shm.c

```diff
@@ -0,0 +1,256 @@
+#include "shm.h"
+#include <sys/mman.h>
+#include <sys/file.h>
+#include <errno.h>
+#include <sys/types.h>
+#include <unistd.h>
+
+
+// Tries to access the coordination shared memory area or creates it if
+// it doesn't yet exist
+mmap_creat_ret_t access_or_create_coord(char* passwd){
+    char* path = malloc(strlen("/ocp_sync")+strlen(passwd)+1);
+    strcpy(path, "/ocp_sync");
+    strcat(path, passwd);
+    int fdm = shm_open(path, O_CREAT | O_EXCL | O_RDWR | O_TRUNC, 0660);
+    printf("opn_sync %d %d\n", fdm, errno);
+    if (fdm == -1){
+        if (errno == EEXIST){
+            fdm = shm_open(path, O_RDWR, 0);
+            coord_struct* str = (coord_struct*) mmap(NULL, sizeof(coord_struct), PROT_READ | PROT_WRITE, MAP_SHARED, fdm, 0);
+            printf("mmp sync %p %d\n", str, errno);
+            return (mmap_creat_ret_t){str, fdm, path};
+        } else {
+            err_and_leave("Error when accessing shared memory", 5);
+        }
+    }
+    ftruncate(fdm, sizeof(coord_struct));
+    coord_struct* str = /*malloc(sizeof(coord_struct));/*/ (coord_struct*) mmap(NULL, sizeof(coord_struct), PROT_READ | PROT_WRITE, MAP_SHARED, fdm, 0);
+    // TODO: init all locks and the like here, returning will relinquish the lock and others might access
+    init_mutex(&(str->lock));
+    init_cond_pack(&(str->reader_ready));
+    init_cond_pack(&(str->writer_ready));
+    str->abort = 0;
+
+    return (mmap_creat_ret_t){str, fdm, path};
```

Comment:
I considered changing to a more verbose struct declaration but since C lacks RAII primitives I think keeping the initialization in a single line is clearer
Definitely to be changed in the conversion to C++